src/shared/services/keys-store.service.ts

```diff
@@ -0,0 +1,162 @@
+ * 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
+ * In applying this license, CERN does not
+ * waive the privileges and immunities granted to it by virtue of its status
+ * as an Intergovernmental Organization or submit itself to any jurisdiction.
+*/
+
+import { Injectable } from '@angular/core';
+import { Map, List, OrderedSet, Iterable, Set, Seq } from 'immutable';
+import { ReplaySubject } from 'rxjs/ReplaySubject';
+
+import { JSONSchema } from '../interfaces';
+import { PathUtilService } from './path-util.service';
+
+@Injectable()
+export class KeysStoreService {
+  private keys$Map: { [path: string]: ReplaySubject<OrderedSet<string>> };
+  private keysMap: { [path: string]: OrderedSet<string> };
+
+  constructor(private pathUtilService: PathUtilService) { }
+
+  forPath(path: string): ReplaySubject<OrderedSet<string>> {
+    return this.keys$Map[path];
+  }
+
+  addKey(path: string, key: string, schema: JSONSchema): string {
+    // FIXME: could do O(n) insert instead of O(nlogn) since the set is already sorted.
+    this.keysMap[path] = this.keysMap[path]
+      .add(key)
+      .sort((a, b) => this.compareBySchema(a, b, schema)) as OrderedSet<string>;
+    this.keys$Map[path].next(this.keysMap[path]);
+    let newKeyPath = `${path}${this.pathUtilService.separator}${key}`;
+
+    let keySchema = schema.properties[key];
+    if (keySchema.type === 'object' || keySchema.componentType === 'table-list') {
+      this.buildKeysMapRecursively(Map<string, any>(), keySchema, newKeyPath);
+    }
+
+    return newKeyPath;
+  }
+
+  deleteKey(path: string, key: string) {
+    this.keysMap[path] = this.keysMap[path].delete(key);
+    this.keys$Map[path].next(this.keysMap[path]);
+    let deletedKeyPath = `${path}${this.pathUtilService.separator}${key}`;
+    delete this.keysMap[deletedKeyPath];
+    delete this.keys$Map[deletedKeyPath];
+  }
+
+  buildKeysMap(json: Map<string, any>, schema: JSONSchema) {
+    this.keys$Map = {};
+    this.keysMap = {};
+    this.buildKeysMapRecursively(json, schema, '');
+  }
+
+  private buildKeysMapRecursively(mapOrList: Iterable<string | number, any>, schema: JSONSchema, path: string) {
+    if (schema.type === 'object') {
+      let map = mapOrList as Map<string, any>;
+      let finalKeys = this.buildkeysForObject(path, map, schema);
+
+      // recursive call
+      finalKeys
+        .filter(key => this.isObjectOrArray(schema.properties[key]))
+        .forEach(key => {
+          let nextPath = `${path}${this.pathUtilService.separator}${key}`;
+          this.buildKeysMapRecursively(map.get(key), schema.properties[key], nextPath);
+        });
+    } else if (schema.componentType === 'table-list') {
+      let list = mapOrList as List<Map<string, any>>;
+      this.buildKeysForTableList(path, list, schema);
+      // there is no recursive call for table list items because they aren't expected to have object or object list as property.
+      // FIXME: really?
+    } else {
```

Comment:
the thing is, this fucntion is never called for a primitive type if you checked the recursive calls. So this else is actually executed for `complex-list`